Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm

```diff
@@ -349,29 +379,184 @@ static inline void parseCSSInjectionOptions(NSDictionary *cssInfo, WebExtensionS
     return true;
 }
 
-NSArray* WebExtensionAPIScripting::toWebAPI(Vector<WebExtensionScriptInjectionResultParameters>& parametersVector)
+bool WebExtensionAPIScripting::validateRegisteredScripts(NSArray *scripts, bool isRegisteringScript, NSString **outExceptionString)
 {
-    auto *results = [NSMutableArray arrayWithCapacity:parametersVector.size()];
+    static NSArray<NSString *> *requiredKeys = @[
+        idKey,
+    ];
 
-    for (auto& parameters : parametersVector) {
-        NSMutableDictionary *result = [NSMutableDictionary dictionaryWithCapacity:3];
+    static NSDictionary<NSString *, id> *keyTypes = @{
+        allFramesKey: @YES.class,
+        cssKey: @[ NSString.class ],
+        excludeMatchesKey: @[ NSString.class ],
+        idKey: NSString.class,
+        jsKey: @[ NSString.class ],
+        matchesKey: @[ NSString.class ],
+        persistAcrossSessionsKey: @YES.class,
+        runAtKey: NSString.class,
+        worldKey: NSString.class,
+    };
 
-        if (parameters.result)
-            result[@"result"] = (NSString *)parameters.result.value();
-        else
-            result[@"result"] = NSNull.null;
+    if (![scripts isKindOfClass:NSArray.class]) {
+        *outExceptionString = toErrorString(nil, @"details", @"an array is expected");
+        return false;
+    }
 
-        ASSERT(parameters.frameID);
-        if (parameters.frameID)
-            result[@"frameId"] = @(WebKit::toWebAPI(parameters.frameID.value()));
+    for (NSDictionary *script in scripts) {
+        if (!validateDictionary(script, @"scripts", requiredKeys, keyTypes, outExceptionString))
+            return false;
 
-        if (parameters.error)
-            result[@"error"] = (NSString *)parameters.error.value();
+        NSString *scriptID = script[idKey];
+        if (!scriptID.length) {
+            *outExceptionString = toErrorString(nil, idKey, @"it must not be empty");
+            return false;
+        }
 
-        [results addObject:result];
+        if ([scriptID characterAtIndex:0] == '_') {
+            *outExceptionString = toErrorString(nil, idKey, @"it must not start with '_'");
+            return false;
+        }
+
+        NSArray *matchPatterns = script[matchesKey];
+        if (isRegisteringScript && !matchPatterns.count) {
+            *outExceptionString = toErrorString(nil, matchesKey, @"it must specify at least one match pattern for script with ID '%@'", script[idKey]);
+            return false;
+        }
+
+        if (matchPatterns && !matchPatterns.count) {
+            *outExceptionString = toErrorString(nil, matchesKey, @"it must not be empty");
+            return false;
+        }
+
+        if (isRegisteringScript && !((NSArray *)script[cssKey]).count && !((NSArray *)script[jsKey]).count) {
```

Comment:
Putting these in local vars might be cleaner than the casts, similar to `matchPatterns` and `scriptID`.